src/libexpr/primops/fromTOML.cc

```diff
@@ -1,93 +1,87 @@
-                visit(*(v.listElems()[i] = state.allocValue()), t2->get()[i]);
-        }
+                        auto & v2 = *state.allocAttr(v, state.symbols.create(elem.first));
 
-        // Handle cases like 'a = [[{ a = true }]]', which IMHO should be
-        // parsed as a array containing an array containing a table,
-        // but instead are parsed as an array containing a table array
-        // containing a table.
-        else if (auto t2 = t->as_table_array()) {
-            size_t size = t2->get().size();
-
-            state.mkList(v, size);
-
-            for (size_t j = 0; j < size; ++j)
-                visit(*(v.listElems()[j] = state.allocValue()), t2->get()[j]);
-        }
+                        // TODO: note about creating children here in old code
+                        visit(v2, elem.second);
+                    }
+                }
+                break;;
+            case toml::value_t::array:
+                {
+                    auto array = toml::get<std::vector<toml::value>>(t);
+
+                    size_t size = array.size();
+                    state.mkList(v, size);
+                    for (size_t i = 0; i < size; ++i)
+                        visit(*(v.listElems()[i] = state.allocValue()), array[i]);
+                }
+                break;;
+            case toml::value_t::boolean:
+                mkBool(v, toml::get<bool>(t));
+                break;;
+            case toml::value_t::integer:
+                mkInt(v, toml::get<int>(t));
+                break;;
+            case toml::value_t::floating:
+                mkFloat(v, toml::get<double>(t));
+                break;;
+            case toml::value_t::string:
+                mkString(v, toml::get<std::string>(t));
+                break;;
+            case toml::value_t::local_datetime:
+            case toml::value_t::offset_datetime:
+            case toml::value_t::local_date:
+            case toml::value_t::local_time:
+                // TODO: convert to string?
+                throw std::runtime_error("Dates and times are not supported");
+                break;;
+            case toml::value_t::empty:
+                mkNull(v);
+                break;;
 
-        else if (t->is_value()) {
-            if (auto val = t->as<int64_t>())
-                mkInt(v, val->get());
-            else if (auto val = t->as<NixFloat>())
-                mkFloat(v, val->get());
-            else if (auto val = t->as<bool>())
-                mkBool(v, val->get());
-            else if (auto val = t->as<std::string>())
-                mkString(v, val->get());
-            else
-                throw EvalError("unsupported value type in TOML");
         }
-
-        else abort();
     };
 
     try {
-        visit(v, parser(tomlStream).parse());
-    } catch (std::runtime_error & e) {
+        visit(val, toml::parse(tomlStream)); // TODO give filename
+    } catch (std::exception & e) { // TODO: toml::syntax_error
```

Comment:
`toml11` has its own exception type, happy to use anything here.

Reply:
We should rethrow those exceptions as a subtype of `Error` since we don't really handle `std::exception` in a lot of places.

Reply:
Do you mean that `std::exception` is too general, and instead we should let those bubble up and only catch e.g. `toml::syntax_error`, which is `toml11`'s (`std::runtime_error` inheriting) exception type? If you mean something else I'm happy to implement it, I just need some pointers!